src/net.cpp

```diff
@@ -718,6 +720,127 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
+    } else {
+        const unsigned int remaining = m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN - m_data_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        // extend buffer, respect previous copied AAD part
+        if (vRecv.size() < CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes) {
+            // Allocate up to 256 KiB ahead, but never more than the total message size (incl. AAD & TAG).
+            vRecv.resize(std::min(CHACHA20_POLY1305_AEAD_AAD_LEN + m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN, CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes + 256 * 1024 + CHACHA20_POLY1305_AEAD_TAG_LEN));
+        }
+
+        memcpy(&vRecv[CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos], pch, copy_bytes);
+        m_data_pos += copy_bytes;
+
+        return copy_bytes;
+    }
+}
+
+CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time)
+{
+    // In v2, vRecv contains the encrypted payload plus the MAC tag (1+bytes serialized message command + ? bytes message payload + 16 byte mac tag)
+    assert(Complete());
+
+    // defensive decoding (MAC check, decryption, command deserialization)
+    // we'll always return a CNetMessage (even if encryption fails), we always increase the AEAD sequence numbers
+    bool valid_checksum = false;
+    bool valid_header = false;
+    std::string command_name;
+
+    if (m_aead->Crypt(m_payload_seqnr, m_aad_seqnr, m_aad_pos, (unsigned char*)vRecv.data(), vRecv.size(), (const uint8_t*)vRecv.data(), vRecv.size(), false)) {
+        // MAC check was successful
+        valid_checksum = true;
+
+        // okay, we could decrypt it, now remove packet length and MAC tag
+        assert(vRecv.size() > CHACHA20_POLY1305_AEAD_AAD_LEN + CHACHA20_POLY1305_AEAD_TAG_LEN);
+        // CDataStream::erase at the begin will just increase the read pos
+        vRecv.erase(vRecv.begin(), vRecv.begin() + CHACHA20_POLY1305_AEAD_AAD_LEN);
+        vRecv.erase(vRecv.end() - CHACHA20_POLY1305_AEAD_TAG_LEN, vRecv.end());
+
+        uint8_t size_or_shortid = 0;
+        try {
+            vRecv >> size_or_shortid;
+        } catch (const std::ios_base::failure&) {
+            LogPrint(BCLog::NET, "Invalid command name\n");
+        }
+        if (size_or_shortid > 0 && size_or_shortid <= 12 && vRecv.size() >= size_or_shortid) {
+            // string command
+            valid_header = true;
+
+            // use direct read since we already read the varlens size uint8_t
+            command_name.resize(size_or_shortid);
+            vRecv.read(&command_name[0], size_or_shortid);
+        }
+        // try for short ID
+        if (!valid_header && size_or_shortid >= 0) {
+            valid_header = true;
+            if (!GetCommandFromShortCommandID(size_or_shortid, command_name)) {
+                // unknown-short-id
+                //  results in a valid but unknown message (will be skipped)
```

Comment:
Maybe it would be better to drop unknown message, or at least as soon as we can to avoid some upstream buffer growing unbounded.

Reply:
I think we should tolerate unknown short IDs the same way as we tolerate unknown string commands (which we currently do for future backward compatibility).

Reply:
Right can we tolerate them, i,e not disconnecting peer and still drop them? We won't be able to process them anyway.